hikari/auto_mod.py

```diff
@@ -0,0 +1,230 @@
+
+if typing.TYPE_CHECKING:
+    import datetime
+
+    from hikari import traits
+
+
+class AutoModActionType(int, enums.Enum):
+    """The type of an auto moderation rule action."""
+
+    BLOCK_MESSAGES = 1
+    """Block the content of the triggering message."""
+
+    SEND_ALERT_MESSAGE = 2
+    """Log the triggering user content to a specified channel."""
+
+    TIMEOUT = 3
+    """Timeout the triggering message's author for a specified duration.
+
+    This type can only be set for `KEYWORD` rules and requires the `MODERATE_MEMBERS`
+    permission to use.
+    """
+
+
+@attr_extensions.with_copy
+@attr.define(kw_only=True, weakref_slot=False)
+class PartialAutoModAction:
+    """Base class for an action which is executed when a ruleis triggered."""
+
+    type: AutoModActionType = attr.field()
+    """The type of auto moderation action."""
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class AutoModBlockMessage(PartialAutoModAction):
+    """Block the content of the triggering message."""
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class AutoModSendAlertMessage(PartialAutoModAction):
+    """Log the triggering content to a specific channel."""
+
+    channel_id: snowflakes.Snowflake = attr.field()
+    """ID of the channel to log trigger events to."""
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class AutoModTimeout(PartialAutoModAction):
+    """Timeout the triggering message's author for a specified duration.
+
+    This type can only be set for `KEYWORD` rules and requires the `MODERATE_MEMBERS`
+    permission to use.
+    """
+
+    duration: datetime.timedelta = attr.field()
+    """The total seconds to timeout the user for (max 2419200 seconds/4 weeks)."""
+
+
+class AutoModEventType(int, enums.Enum):
+    """Type of event to check for an auto moderation rule."""
+
+    MESSAGE_SEND = 1
+    """When a member sends or edits a message in the guild."""
+
+
+class AutoModTriggerType(int, enums.Enum):
+    """Type of trigger for an auto moderation rule."""
+
+    KEYWORD = 1
+    """Match message content against a list of keywords."""
+
+    HARMFUL_LINK = 2
+    """Scan messages for links which are deemed "harmful" by Discord."""
+
+    SPAM = 3
+    """Discord's guild anti-spam system."""
+
+    KEYWORD_PRESET = 4
+    """Discord's preset keyword triggers."""
+
+
+class AutoModKeywordPresetType(int, enums.Enum):
+    """Discord's KEYWORD_PRESET type."""
+
+    PROFANITY = 1
+    """Trigger on words which may be considered forms of swearing or cursing."""
+
+    SEXUAL_CONTENT = 2
+    """Trigger on words that refer to explicit behaviour or activity."""
+
+    SLURS = 3
+    """Trigger on personal insults and words which "may be considered hate speech"."""
+
+
+@attr_extensions.with_copy
+@attr.define(kw_only=True, weakref_slot=False)
+class PartialAutoModTrigger:
+    """Base class representing the content a rule triggers on."""
+
+    type: AutoModTriggerType = attr.field(eq=False, hash=False, repr=False)
```

Comment:
Add defaults for this or just have it be static per-sub-class

Reply:
If we never actually get this class in the real world, this could be made abstract and then specify in each subclass